lib/HLSL/HLLowerUDT.cpp

```diff
@@ -285,11 +269,10 @@ void hlsl::ReplaceUsesForLoweredUDT(Value *V, Value *NewV) {
     } else if (AddrSpaceCastInst *AC = dyn_cast<AddrSpaceCastInst>(user)) {
       // Address space cast
       IRBuilder<> Builder(AC);
-      ReplaceUsesForLoweredUDT(
-          user, InsertAddrSpaceCastIfRequired(
-                    Builder, NewV, AC->getType()->getPointerAddressSpace()));
+      Value *NewAC = Builder.CreateAddrSpaceCast(
+          NewV, PointerType::get(Ty, AC->getType()->getPointerAddressSpace()));
+      ReplaceUsesForLoweredUDT(user, NewAC);
```

Comment:
One point of `InsertAddrSpaceCastIfRequired` was to avoid inserting the cast when *not* required.  I'm still not sure we have the testing to cover the scenario properly.

Reply:
IRBuilder::CreateAddrSpaceCast will just return the input value if the test matches.

Reply:
I think Xiang is referring to this code: 

https://github.com/microsoft/DirectXShaderCompiler/blob/52ace1c57cb782f02ae8ae7fe011bf9c5258a295/include/llvm/IR/IRBuilder.h#L1331

It does look like it will return the right thing when the types fully match.